### PR TITLE
Add more options for generating bucket keys

### PR DESCRIPTION
**Pull request by szhengac**
## Description ##
This PR adds feature supporting bucket keys with linear and exponential increasing bucket width. 

In my testing on WMT16, with the linear and exponential schemes, the throughput improves 16% and 18%, respectively. On IWSLT2015, I observe a 92% increase on throughput. 

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by mli** [comment]
Job PR-145/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-145/3/index.html

**Comment by mli** [comment]
Job PR-145/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-145/4/index.html

File: gluonnlp/data/sampler.py [hunk 1]
```diff
@@ -63,6 +65,150 @@ def _bucket_average_lengths(bucket_sample_ids, seq_lengths):
     return bucket_average_lengths
 
 
+class Bucket(object):
```
Comment (sxjscience) [inline gluonnlp/data/sampler.py]: How about renaming it to `BucketGen` ?

File: gluonnlp/data/sampler.py [hunk 2]
```diff
@@ -63,6 +65,150 @@ def _bucket_average_lengths(bucket_sample_ids, seq_lengths):
...
+class ConstWidthBucket(Bucket):
+    r""" Buckets with constant width.
+    """
+    def __call__(self, single_element, max_lengths, min_lengths, num_buckets):
+        r"""This generate bucket keys given that all the buckets have the same width.
+
+        Parameters
+        ----------
+        single_element : bool
+            Whether lengths of sequences is a list of int.
```
Comment (sxjscience) [inline gluonnlp/data/sampler.py]: It should be possible to remove the `single_element` argument.

File: gluonnlp/data/sampler.py (hunk 1 above)
Comment (sxjscience) [reply gluonnlp/data/sampler.py]: or `BucketScheme`

File: gluonnlp/data/sampler.py [hunk 3]
```diff
@@ -120,7 +266,11 @@ class FixedBucketSampler(Sampler):
         False: each batch contains batch_size sequences, number of sequence elements varies.
         True: each batch contains batch_size elements, number of sequences varies. In this case,
         ratio option is ignored.
-
+    bucket : Bucket, default ConstWidthBucket
+        It is used to generate bucket keys. It supports:
+        ConstWidthBucket: all the buckets have the same width
+        LinearWidthBucket: the width of ith  bucket follows :math:`w_i = alpha * i + 1`
```
Comment (sxjscience) [inline gluonnlp/data/sampler.py]: The alpha here should be `\alpha`.

File: gluonnlp/data/sampler.py [hunk 4]
```diff
@@ -63,6 +65,150 @@ def _bucket_average_lengths(bucket_sample_ids, seq_lengths):
...
+    Parameters
+    ----------
+    bucket_len_step : float, default 1.1
+        This is the increasing factor for the bucket width.
+    """
+    def __init__(self, bucket_len_step=1.1):
+        self.bucket_len_step = bucket_len_step
+
+    def __call__(self, single_element, max_lengths, min_lengths, num_buckets):
+        r"""This function implements a scheme for exponentially increasig bucket width:
```
Comment (sxjscience) [inline gluonnlp/data/sampler.py]: ... increasing ...

**Comment by mli** [comment]
Job PR-145/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-145/6/index.html

File: gluonnlp/data/sampler.py [hunk 5]
```diff
@@ -63,6 +65,150 @@ def _bucket_average_lengths(bucket_sample_ids, seq_lengths):
...
+                 for i in range(num_buckets)]
+            bucket_keys[-1] = tuple(max(max_bucket_key, max_len)
+                                    for max_bucket_key, max_len
+                                    in zip(bucket_keys[-1], max_lengths))
+        else:
+            alpha = 2 * float(max_lengths - min_lengths - num_buckets) \
+                    / (num_buckets * (num_buckets + 1))
+            bucket_keys = [int(round(min_lengths + alpha * (((i + 1) * (i + 2)) / 2) + i + 1))
+                           for i in range(num_buckets)]
+            bucket_keys[-1] = max(bucket_keys[-1], max_lengths)
```
Comment (sxjscience) [inline gluonnlp/data/sampler.py]: Should we apply the filter `max(ele, max_lengths)` to all the bucket_keys?

Comment (szhengac) [reply gluonnlp/data/sampler.py]: No, we only need to make sure the last one is at least equal to max_lengths. If it is applied to all the bucket keys, then every key will equal max_lengths.

Comment (sxjscience) [reply gluonnlp/data/sampler.py]: I see. I was considering the case when the bucket_keys could be larger than the max_lengths. So we  will waste some resources.

Comment (szhengac) [reply gluonnlp/data/sampler.py]: Theoretically, the difference will be within 1.

**Comment by mli** [comment]
Job PR-145/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-145/7/index.html

File: gluonnlp/data/sampler.py [hunk 6]
```diff
@@ -63,6 +66,141 @@ def _bucket_average_lengths(bucket_sample_ids, seq_lengths):
     return bucket_average_lengths
 
 
+class BucketScheme(object):
+    r"""Base class of for generating bucket keys.
```
Comment (sxjscience) [inline gluonnlp/data/sampler.py]: docstring

Comment (sxjscience) [reply gluonnlp/data/sampler.py]: ... "of for" ...

**Comment by mli** [comment]
Job PR-145/8 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-145/8/index.html

**Comment by mli** [comment]
Job PR-145/9 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-145/9/index.html

**Comment by mli** [comment]
Job PR-145/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-145/10/index.html

File: docs/api/data.rst [hunk 7]
```diff
@@ -135,6 +135,15 @@ iterate through sequence data.
     FixedBucketSampler
     SortedBucketSampler
 
+The `FixedBucketSampler` uses following bucket classes to generate bucket keys.
```
Comment (szha) [inline docs/api/data.rst]: bucketing scheme classes

File: gluonnlp/_constants.py [hunk 8]
```diff
@@ -36,6 +38,9 @@
 
 LARGE_NEGATIVE_FLOAT = -LARGE_POSITIVE_FLOAT
 
+INT_TYPES = (int, np.int8, np.int, np.int16, np.int32, np.int64,
+             np.uint8, np.uint16, np.uint32, np.uint64)
```
Comment (szha) [inline gluonnlp/_constants.py]: Do we need all these options? Also, we can probably reuse what has been declared in mxnet base. https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/base.py#L40

File: gluonnlp/data/sampler.py [hunk 9]
```diff
@@ -82,6 +79,141 @@ def _bucket_average_lengths(bucket_sample_ids, seq_lengths):
...
+                           for i in range(num_buckets)]
+        return bucket_keys
+
+
+class LinearWidthBucket(BucketScheme):
+    r""" Buckets with linearly increasing width.
+    """
+    def __call__(self, max_lengths, min_lengths, num_buckets):
+        r"""This implements a scheme for linearly increasing bucket width:
+                :math:`w_i = \alpha * i + 1` for all :math:`i \geq 1`.
```
Comment (szha) [inline gluonnlp/data/sampler.py]: put this in class doc. method doc should explain what the method's arguments and return values are

File: gluonnlp/data/sampler.py [hunk 10]
```diff
@@ -82,6 +79,141 @@ def _bucket_average_lengths(bucket_sample_ids, seq_lengths):
...
+    ----------
+    bucket_len_step : float, default 1.1
+        This is the increasing factor for the bucket width.
+    """
+    def __init__(self, bucket_len_step=1.1):
+        self.bucket_len_step = bucket_len_step
+
+    def __call__(self, max_lengths, min_lengths, num_buckets):
+        r"""This function implements a scheme for exponentially increasing bucket width:
+                :math:`w_i = bucket_len_step * w_{i-1}` for all :math:`i \geq 2`.
```
Comment (szha) [inline gluonnlp/data/sampler.py]: same

File: gluonnlp/_constants.py (hunk 8 above)
Comment (szhengac) [reply gluonnlp/_constants.py]: You are right. We may not really need something like int8.

**Comment by mli** [comment]
Job PR-145/13 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-145/13/index.html

File: gluonnlp/data/sampler.py [hunk 11]
```diff
@@ -82,6 +79,141 @@ def _bucket_average_lengths(bucket_sample_ids, seq_lengths):
...
+
+        Parameters
+        ----------
+        max_lengths : int of list of int
+            Maximum of lengths of sequences.
+        min_lengths : int of list of int
+            Minimum of lengths of sequences.
+        num_buckets : int
+            Number of buckets
+        """
```
Comment (szha) [inline gluonnlp/data/sampler.py]: describe the type of output

**Comment by mli** [comment]
Job PR-145/14 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-145/14/index.html

**Pull request merged**